### PR TITLE
Add VideoFrameDecoder to support decoding video frames from any source.

### LINKED ISSUE #687: Support loading video frames from the network.
**Is your feature request related to a problem? Please describe.**
Currently, Coil only supports decoding video frames from local sources (files or local URIs). 

**Describe the solution you'd like**
Coil should support decoding video frames from a URL.

**Additional context**
This is something that I'd really like to add! Unfortunately, the API Coil uses to decode video frames, [`MediaMetadataRetriever`](https://developer.android.com/reference/android/media/MediaMetadataRetriever), has limitations 
around using a network-backed `Source` and works much better with a `File`. A Coil `Decoder` intentionally has no knowledge of `File`s to allow the source of an image to be completely decoupled from decoding logic.

### PR DESCRIPTION
**Pull request by colinrtwhite**
Fixes #687.

This finally adds support for decoding video frame from any source (including the network!) with an important caveat. `VideoFrameDecoder` creates a short lived file on disk to buffer the source into. This is less performant than decoding the file in memory and it requires enough space on disk to copy the video which could be a problem for very large video files.

`VideoFrameFileFetcher` and `VideoFrameUriFetcher` don't have this caveat which is why they're not deprecated. Long term (possibly in the next major version) we'll hopefully be able to remove this limitation on `VideoFrameDecoder` and deprecate the fetchers.

### CODE REVIEW

File: coil-video/src/main/java/coil/decode/VideoFrameDecoderDelegate.kt [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import androidx.core.graphics.drawable.toDrawable
+import coil.bitmap.BitmapPool
+import coil.request.videoFrameMicros
+import coil.request.videoFrameOption
+import coil.size.OriginalSize
+import coil.size.PixelSize
+import coil.size.Size
+import kotlin.math.roundToInt
+
+internal class VideoFrameDecoderDelegate(private val context: Context) {
```
Comment (colinrtwhite) [inline coil-video/src/main/java/coil/decode/VideoFrameDecoderDelegate.kt:22]: This class is basically pulled out entirely from `VideoFrameFetcher` without changes.

File: coil-video/src/main/java/coil/decode/VideoFrameDecoder.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        return mimeType != null && mimeType.startsWith("video/")
+    }
+
+    override suspend fun decode(
+        pool: BitmapPool,
+        source: BufferedSource,
+        size: Size,
+        options: Options
+    ): DecodeResult {
+        val tempFile = File.createTempFile("tmp", null, context.cacheDir.apply { mkdirs() })
```
Comment (Jawnnypoo) [inline coil-video/src/main/java/coil/decode/VideoFrameDecoder.kt:33]: We don't need write permissions to do this, right?

**Review by Jawnnypoo — APPROVED** [review]
👍 Great work, I know this is much appreciated from those who need it.

File: coil-video/src/main/java/coil/decode/VideoFrameDecoder.kt (hunk 2 above)
Comment (colinrtwhite) [reply coil-video/src/main/java/coil/decode/VideoFrameDecoder.kt:33]: Yep! The `cacheDir` is free to write to.

**Comment by ultraon** [comment]
It doesn't look like a workable solution in conjunction with streaming content urls (mpd - MPEG DASH, m3u8 - HLS), right?

**Comment by colinrtwhite** [comment]
@ultraon It should work with [any format supported by Android](https://developer.android.com/guide/topics/media/media-formats#video-codecs).

**Comment by yehia2030** [comment]
it still didn't work with me in compose and this is my example

            val context = LocalContext.current
            Image(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(200.dp)
                    .padding(horizontal = 10.dp),
                painter = rememberAsyncImagePainter(ImageRequest.Builder(LocalContext.current)
                    .data(data = "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4")
                    .apply(block = {
                     decoderFactory(VideoFrameDecoder.Factory())
                    }).build()),
                contentDescription = "test",
            )

if there is any solution please let me know

**Comment by yehia2030** [comment]
after  a lot of pain and i got a black image 

            val imageLoader = ImageLoader.Builder(context).build()
            val model = ImageRequest.Builder(LocalContext.current)
                .data(data = "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4")
                .apply(block = {
                    decoderFactory(VideoFrameDecoder.Factory())
                    videoFrameMillis(50)
                    listener(
                        onStart = {println("Start")},
                        onError = { _, result -> println("Error ${result.throwable.message}")},
                        onSuccess = { _, _ ->  println("Success")},
                        onCancel = { println("Cancel") }
                    )
                    size(100)
                }).build()
            Image(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(200.dp)
                    .padding(horizontal = 10.dp),
                painter = rememberAsyncImagePainter(model, imageLoader = imageLoader,
                    onError = {
                    println("Error 2 ${it.result.throwable.message}")
                }, onLoading = {
                        println("Loading 2 ")
                    }, onSuccess = {
                        println("Success 2${it.result.isSampled} ")

                    }),
                contentDescription = "test"
            )
            LaunchedEffect(key1 = true) {
                imageLoader.execute(model)
            }

**Comment by yehia2030** [comment]
could we make a buffer of something like that as if the video is long it takes too much time to load the image  ? as i think it loaded the fully video then get its thumbnail

**Pull request merged**